clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerProtocol.java

```diff
@@ -102,53 +106,97 @@
+                // for any versions higher than known, try to serialize it as V1
+                return serializeSubscriptionV1(subscription);
         }
+    }
 
-        public short code() {
-            return code;
+    public static Subscription deserializeSubscription(ByteBuffer buffer) {
+        Struct header = CONSUMER_PROTOCOL_HEADER_SCHEMA.read(buffer);
+        Short version = header.getShort(VERSION_KEY_NAME);
+
+        if (version < CONSUMER_PROTOCOL_V0)
+            throw new SchemaException("Unsupported subscription version: " + version);
+
+        switch (version) {
+            case CONSUMER_PROTOCOL_V0:
+                return deserializeSubscriptionV0(buffer);
+
+            case CONSUMER_PROTOCOL_V1:
+                return deserializeSubscriptionV1(buffer);
+
+            // assume all higher versions can be parsed as V1
+            default:
+                return deserializeSubscriptionV1(buffer);
         }
+    }
 
-        public static AssignmentError fromCode(final short code) {
-            switch (code) {
-                case 0:
-                    return NONE;
-                case 1:
-                    return NEED_REJOIN;
-                default:
-                    throw new IllegalArgumentException("Unknown error code: " + code);
-            }
+    public static ByteBuffer serializeAssignment(Assignment assignment, Short version) {
+        switch (version) {
+            case CONSUMER_PROTOCOL_V0:
+                return serializeAssignmentV0(assignment);
+
+            case CONSUMER_PROTOCOL_V1:
+                return serializeAssignmentV1(assignment);
+
+            default:
+                // for any versions higher than known, try to serialize it as V1
+                return serializeAssignmentV1(assignment);
+        }
+    }
+
+    public static Assignment deserializeAssignment(ByteBuffer buffer) {
+        Struct header = CONSUMER_PROTOCOL_HEADER_SCHEMA.read(buffer);
+        Short version = header.getShort(VERSION_KEY_NAME);
+
+        if (version < CONSUMER_PROTOCOL_V0)
+            throw new SchemaException("Unsupported assignment version: " + version);
+
+        switch (version) {
+            case CONSUMER_PROTOCOL_V0:
+                return deserializeAssignmentV0(buffer);
+
+            case CONSUMER_PROTOCOL_V1:
+                return deserializeAssignmentV1(buffer);
+
+            default:
+                // assume all higher versions can be parsed as V1
+                return deserializeAssignmentV1(buffer);
         }
     }
 
-    public static ByteBuffer serializeSubscriptionV0(PartitionAssignor.Subscription subscription) {
+    static ByteBuffer serializeSubscriptionV0(Subscription subscription) {
         Struct struct = new Struct(SUBSCRIPTION_V0);
         struct.set(USER_DATA_KEY_NAME, subscription.userData());
-        struct.set(TOPICS_KEY_NAME, subscription.topics().toArray());
+
+        ConsumerSubscriptionData consumerData = subscription.consumerData();
```

Comment:
Note the `struct.set` calls does not need to be in order, as the schema definition's ordering is the source of truth.